pkg/controller/messagingproject/messagingproject_controller.go

```diff
@@ -362,37 +431,144 @@ func (r *ReconcileMessagingProject) Reconcile(request reconcile.Request) (reconc
 	return result.Result(), err
 }
 
-func findBestMatch(project *v1.MessagingProject, infras []v1.MessagingInfrastructure) *v1.MessagingInfrastructure {
-	var bestMatch *v1.MessagingInfrastructure
-	var bestMatchSelector *v1.NamespaceSelector
-	for _, infra := range infras {
-		if infra.Status.Phase != v1.MessagingInfrastructureActive {
-			continue
+/**
+ * Checks if a selector matches a given project.
+ */
+func (r *ReconcileMessagingProject) matchesSelector(ctx context.Context, namespace string, selector *v1.NamespaceSelector) (bool, error) {
+	if selector == nil {
+		return true, nil
+	}
+
+	for _, ns := range selector.MatchNames {
+		if ns == namespace {
+			return true, nil
 		}
-		selector := infra.Spec.NamespaceSelector
+	}
+
+	// Fall back to looking up namespace to verify labels if we have specified. This ensures that we only require namespace lookup permissions if
+	// absolutely necessary.
+	if len(selector.MatchLabels) > 0 || len(selector.MatchExpressions) > 0 {
+		ns := &corev1.Namespace{}
+		err := r.client.Get(ctx, types.NamespacedName{Name: namespace}, ns)
+		if err != nil {
+			return false, err
+		}
+
+		if len(selector.MatchLabels) > 0 {
+			sel, err := metav1.LabelSelectorAsSelector(&metav1.LabelSelector{
+				MatchLabels: selector.MatchLabels,
+			})
+			if err != nil {
+				return false, err
+			}
+			if sel.Matches(labels.FromMap(ns.Labels)) {
+				return true, nil
+			}
+		}
+
+		if len(selector.MatchExpressions) > 0 {
+			sel, err := metav1.LabelSelectorAsSelector(&metav1.LabelSelector{
+				MatchExpressions: selector.MatchExpressions,
+			})
+			if err != nil {
+				return false, err
+			}
+			return sel.Matches(labels.FromMap(ns.Labels)), nil
+		}
+	}
+	return false, nil
+}
+
+/**
+ * Find the best matching object among a list of objects whose selectors are matched against a messaging project.
+ */
+func (r *ReconcileMessagingProject) findBestMatch(ctx context.Context, namespace string, objects []v1.Selectable) (v1.Selectable, error) {
+	var bestMatch v1.Selectable
+	var bestMatchSelector *v1.NamespaceSelector
+
+	// Try to find a selector matching a specific name before needing to lookup namespace
+	for _, object := range objects {
+		selector := object.GetSelector()
 		// If there is a global one without a selector, use it
 		if selector == nil && bestMatch == nil {
-			bestMatch = &infra
+			bestMatch = object
```

Comment:
I don't think there should be more than one global on, but what happens if there are?

Reply:
It would just select the first one. I can add so that it orders by creation date which is the resolution used for the case where multiple infras match on the labels or names.

Reply:
I think ordering by creation date would be best from consistency point of view.

Reply:
Done, also added tests for this.